dm/dm-replication-logic.md

```diff
@@ -0,0 +1,153 @@
+
+## DML processing flow
+
+The Sync unit processes DML statements as follows:
+
+1. Read the binlog event from the MySQL, MariaDB, or relay log.
+2. Transform the binlog event read from the data source:
+
+    1. [Binlog filter](dm/dm-binlog-event-filter.md): Filter binlog events according to binlog expressions, configured by `filters`.
+    2. [Table routing](dm/dm-table-routing.md): Transform the "database/table" name according to the "database/table" routing rule, configured by `routes`.
+    3. [Expression filter](/filter-dml-event.md): Filter binlog events according to SQL expressions, configured by `expression-filter`.
+
+3. Optimize the DML execution plan:
+
+    1. [Compactor](#compactor): Merge multiple operations on the same record (with the same primary key) into one operation. This feature is enabled by `syncer.compact`.
+    2. [Causality](#causality): Perform conflict detection on different records (with different primary keys) to improve the concurrency of replication.
+    3. [Merger](#merger): Merge multiple binlog events into one DML statement, enabled by `syncer.multiple-rows`.
+
+4. Execute the DML to the downstream.
+5. Periodically save the binlog position or GTID to the checkpoint.
+
+![DML processing logic](/media/dm/dm-dml-replication-logic.png)
+
+## DML optimization logic
+
+The Sync unit implements the DML optimization logic through three steps: Compactor, Causality, and Merger.
+
+### Compactor
+
+According to the binlog records of the upstream, DM captures the changes of the records and replicates them to the downstream. When the upstream makes multiple changes to the same record (`INSERT`/`UPDATE`/`DELETE`) in a short period of time, DM can compress multiple changes into one change through Compactor to reduce the pressure on the downstream and improve the throughput. For example:
+
+```
+INSERT + UPDATE => INSERT
+INSERT + DELETE => DELETE
+UPDATE + UPDATE => UPDATE
+UPDATE + DELETE => DELETE
+DELETE + INSERT => UPDATE
+```
+
+The Compactor feature is disabled by default. To enable it, you can set `syncer.compact` to `true` in the `sync` configuration module of the replication task, as shown below:
+
+```yaml
+syncers:                            # The configuration parameters of the sync processing unit
+  global:                           # Configuration name
+    ...                              # Other configurations are omitted
+    compact: true
+```
+
+### Causality
+
+The sequential synchronization model of MySQL binlog requires that binlog events be synchronized in the order of binlog. This synchronization model cannot meet the requirements of high QPS and low synchronization latency. In addition, because not all operations involved in binlog have conflicts, sequential synchronization is not necessary in those cases.
```

Comment:
```suggestion
The sequential replication model of MySQL binlog requires that binlog events be replicated in the order of binlog. This replication model cannot meet the requirements of high QPS and low replication latency. In addition, because not all operations involved in binlog have conflicts, sequential replication is not necessary in those cases.
```